core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.coordinator.transaction
+
+import java.util
+import java.util.concurrent.LinkedBlockingQueue
+
+import kafka.common.{BrokerEndPointNotAvailableException, RequestAndCompletionHandler}
+import kafka.server.{DelayedOperationPurgatory, MetadataCache}
+import kafka.utils.Logging
+import org.apache.kafka.common.network.ListenerName
+import org.apache.kafka.common.requests.{TransactionResult, WriteTxnMarkersRequest}
+import org.apache.kafka.common.requests.WriteTxnMarkersRequest.TxnMarkerEntry
+import org.apache.kafka.common.utils.{Time, Utils}
+import org.apache.kafka.common.{Node, TopicPartition}
+
+import scala.collection.{concurrent, immutable, mutable}
+import collection.JavaConverters._
+import collection.JavaConversions._
+
+class TransactionMarkerChannel(interBrokerListenerName: ListenerName,
+                               metadataCache: MetadataCache,
+                               time: Time = Time.SYSTEM) extends Logging {
+
+  // we need the metadataPartition so we can clean up when Transaction Log partitions emigrate
+  case class PendingTxnKey(metadataPartition: Int, producerId: Long)
+
+  private val brokerStateMap: concurrent.Map[Int, DestinationBrokerAndQueuedMarkers] = concurrent.TrieMap.empty[Int, DestinationBrokerAndQueuedMarkers]
+  private val pendingTxnMap: concurrent.Map[PendingTxnKey, TransactionMetadata] = concurrent.TrieMap.empty[PendingTxnKey, TransactionMetadata]
+
+  // TODO: What is reasonable for this
+  private val brokerNotAliveBackoffMs = 10
+
+  // visible for testing
+  private[transaction] def queueForBroker(brokerId: Int) = {
+    brokerStateMap.get(brokerId)
+  }
+
+  private[transaction]
+  def drainQueuedTransactionMarkers(txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker]): Iterable[RequestAndCompletionHandler] = {
```

Comment:
@dguy @junrao What's the motivation of trying to drain all the queued elements? Since the max inflight request is only 1 in the network client, even if we construct multiple requests for a certain destination only the first request will succeed in sending right? In that case could just do the 1) peek-first 2) if-ready-send-and-pop pattern?